event-web-framework/draft.markdown

```diff
@@ -0,0 +1,872 @@
+4. Otherwise, we check if the buffer is one of `complete?`, `too-big?`, `too-old?` or `too-needy?`. If it's any of them, we remove it from the connections table and return the appropriate HTTP response.
+
+This is the first time we're seeing I/O in our event loop. In our discussion in FIXME SECTIONREF, we mentioned that we have to be very careful about I/O in an event-driven system, because we could accidentally block our single thread. So, what do we do here to ensure that this doesn't happen? We have to explore our implementation of `buffer!` to find out exactly how this works.
+
+### Processing Connections Without Blocking
+
+The basis of our approach to processing connections without blocking is the library function [`read-char-no-hang`](http://clhs.lisp.se/Body/f_rd_c_1.htm), which immediately returns `nil` when called on a stream that has no available data. Where there is data to be read, we use a buffer to store intermediate input for this connection.
+
+```lisp
+    (defmethod buffer! ((buffer buffer))
+      (handler-case
+          (let ((stream (bi-stream buffer)))
+    	(incf (tries buffer))
+    	(loop for char = (read-char-no-hang stream) until (null char)
+    	   do (push char (contents buffer))
+    	   do (incf (total-buffered buffer))
+    	   when (request buffer) do (decf (expecting buffer))
+    	   when (line-terminated? (contents buffer))
+    	   do (multiple-value-bind (parsed expecting) (parse buffer)
+    		(setf (request buffer) parsed
+    		      (expecting buffer) expecting)
+    		(return char))
+    	   when (> (total-buffered buffer) +max-request-size+) return char
+    	   finally (return char)))
+        (error () :eof)))
+```
+
+When `buffer!` is called on a `buffer`, it:
+- increments the `tries` count, so that we can evict "needy" buffers in `process-ready`;
+- loops to read characters from the input stream, and 
+- returns the last character it read if it has read all of the available input. 
+- It also tracks any `\r\n\r\n` sequences so that we can later detect complete requests. 
+- Finally, any error results it returns an `:eof` to signal that `process-ready` should discard this particular connection.
+
+The `buffer` type is a CLOS _class_. Classes in CLOS let us define a type with fields called `slots`. We don't see the behaviours associated with `buffer` on the class definition, because (as we've already learned), we do that using generic functions like `buffer!`.
+
+`defclass` does allow us to specify getters/setters (`reader`s/`accessor`s), and slot initializers; `:initform` specifies a default value, while `:initarg` identifies a hook that the caller of `make-instance` can use to provide a default value. 
+
+NOTE: Generic functions can specialize on types that aren't classes, right? The example of `port` from the first method definition didn't seem like it was a class. I get the impression that classes are just giving us some syntactic sugar over common operations on structs.
```

Comment:
Sort of. We have `struct`s in CL too; the main difference between a `class` and a `struct` is that `struct`s don't implement any kind of inheritance. You can see an empty list following the class-name of each definition in this chapter; that can designate a list of superclasses for the class we're defining. For the purposes of this chapter, they're interchangeable (I don't think I ever inherit from anything).

Generic functions can specialize on all of them; `class`es, `struct`s and primitive types like `integer`/`list`/`character`/`string`/etc.

I'm not sure how much of this stuff is worth expanding on inline; but it might be a good idea to refer to more in-depth treatment (Chapters [16](http://www.gigamonkeys.com/book/object-reorientation-generic-functions.html) and [17](http://www.gigamonkeys.com/book/object-reorientation-classes.html) of PCL, for example).